DRAFT: CLE specification (#3)
Ecma-TC54/tg3: SPECIFICATION.md

# Common Lifecycle Enumeration (CLE) Specification v1.0.0

## Overview

The Common Lifecycle Enumeration (CLE) provides a standardized format for communicating software component lifecycle events in a machine-readable format. This specification defines the JSON schema and requirements for CLE documents.

## Requirements Terminology

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://datatracker.ietf.org/doc/html/rfc2119).

## Schema Definition

### Top-level Fields

**Required Fields:**

| Field | Type | Description |
|-------|------|-------------|

> | `events` | array | Ordered array of Event objects representing the component's lifecycle events. MUST be chronologically ordered by effective date. |

Author: dont use CAPS

### Event Types

#### generalAvailability
*Category: Version Event*

Indicates when a component version is released and available for use.

**Additional Required Fields:**
- `version`: VERS-formatted single version (e.g., "vers:npm/1.0.0")

Example:
```json
{
  "type": "generalAvailability",
  "effective": "2023-01-01T00:00:00Z",
  "version": "vers:npm/1.0.0",
  "published": "2023-01-01T00:00:00Z",
  "modified": "2023-01-01T00:00:00Z",
  "description": "Version 1.0.0 is now generally available",
  "references": [
    "https://example.com/release-notes"
  ]
}
```

#### endOfDevelopment
*Category: Version Event*

Indicates when the manufacturer ceases development of a component or service.

**Additional Required Fields:**
- `range`: VERS-formatted version range (e.g., "vers:npm/>=1.0.0|<2.0.0")

Example:
```json
{
  "type": "endOfDevelopment",
  "effective": "2024-01-01T00:00:00Z",
  "range": "vers:npm/>=1.0.0|<2.0.0",
  "published": "2023-06-01T00:00:00Z",
  "modified": "2023-06-01T00:00:00Z",
  "description": "Development of 1.x versions will cease"
}
```

#### endOfSupport
*Category: Version Event*

> Indicates when the manufacturer ceases any and all support of a component or service. This point in time marks a transfer of risk from the manufacturer to the consuming organisation or user of the component or service, encompassing all cybersecurity knowledge and known vulnerabilities, with no further assistance provided by the manufacturer.

Reviewer: This has probably been discussed before, but if `endOfSupport` means that the manufacturer stops accepting vulnerability reports and will not publish any new CVEs, how do you mark the "end of security updates"? There is a certain number of vulnerability reports, which are "UNSUPPORTED WHEN ASSIGNED". These are published, but the manufacturer will never fix them.

Maybe `endOfSupport` should not exonerate the manufacturer from accepting vulnerability reports, which would be published until `endOfLife` is announced.

Author: > how do you mark the "end of security updates"

`endOfSupport` is the way for doing that since it  "indicates when the manufacturer ceases any and all support of a component or service"

`endOfSupport` should not care whether the manufacturer accepts vulnerabilities reports or not, just that they will not be doing any new releases to fix known vulnerabilities.

Reviewer: I am not a native speaker, but from the description I would think that `endOfSupport` also means that vulnerability reports will not be accepted and processed.

Reviewer 2: "end of support" to me definitely sounds like it should be equivalent to, "every maintainer is dead".

Reviewer: I asked a little bit around and most projects don't make a difference between `endOfSupport` and `endOfLife`: terms like EOL and "is no longer supported" are used interchangeably (e.g. [Tomcat's EOLs](https://tomcat.apache.org/tomcat-85-eol.html)).

[SAP LeanIX](https://www.leanix.net/en/wiki/trm/what-is-end-of-life-vs-end-of-support) seems to be one of the few resources that differentiates the terms, although the difference seems mostly symbolic. For both EOS and EOL the manufacturer stops all support, but for EOS the product might "still function".

On the other hand some projects allow some level of support after EOL. For example [HttpCore 4.4.x](https://hc.apache.org/status.html) is EOL but "will be receiving fixes for major defects and security issues only". Should this be `endOfDevelopment` instead?

# Common Lifecycle Enumeration (CLE) Specification v1.0.0

## Overview

The Common Lifecycle Enumeration (CLE) provides a standardized format for communicating software component lifecycle events in a machine-readable format. This specification defines the JSON schema and requirements for CLE documents.

## Requirements Terminology

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://datatracker.ietf.org/doc/html/rfc2119).

## Schema Definition

### Top-level Fields

**Required Fields:**

| Field | Type | Description |
|-------|------|-------------|
| `$schema` | string | URI identifying the [JSON Schema](https://json-schema.org/) document that describes the version of the CLE schema to use (e.g., "https://TODO/cle.v1.0.0.json") |

> | `events` | array | Ordered array of Event objects representing the component's lifecycle events. MUST be chronologically ordered by effective date. |

Reviewer 2: descending or ascending?

Reviewer 2: also, are there any uniqueness guarantees here?

Author: `descending` makes sense to me.

uniqueness is a good question. What happens if you have an event of type `generalAvailability` that covers the same version but has a different `effectiveDate`.

Should we make events unique according to type and version (computed from range if needed)?

Reviewer 2: i mean, it's totally reasonable to issue a new event that overwrites or invalidates a previous one - but there shouldn't ever be two objects that represent the same event.

Reviewer 2: ```suggestion
| `events` | array | Ordered array of Event objects representing the component's lifecycle events. MUST be chronologically ordered by effective date, descending. |
```

### Event Object

The base object that represents a discrete lifecycle event. All events share these common fields.

**Required Fields:**

| Field | Type | Description |
|-------|------|-------------|
| `type` | string | The type of lifecycle event. MUST be one of the defined Event Types. |
| `effective` | string | The time when the event takes effect, as an RFC3339-formatted timestamp in UTC (ending in "Z"). |
| `published` | string | The time when the event was first published, as an RFC3339-formatted timestamp in UTC (ending in "Z"). |
| `modified` | string | The time when the event was last modified, as an RFC3339-formatted timestamp in UTC (ending in "Z"). |

### Event Types

> #### generalAvailability

Reviewer 2: i'm confused by "general". what would be non-general availability?

Author: good question, I suppose if we're specifying the version string it could include prelease versions, so in that case, `released` or `available` is simpler and make more sense?

Reviewer 2: If there's a version number, it's both released and available (and yes, prereleases must be supported)

*Category: Version Event*

Indicates when a component version is released and available for use.

**Additional Required Fields:**

| Field | Type | Description |
|-------|------|-------------|

> | `versions` | array[string] | Array of simple version strings (e.g., ["1.0.0", "1.1.0"]). Required for version-specific events. |

Reviewer 2: is there a VERS standard for single versions? if so, these should be required to be there.

also i'm not sure why we'd want both versions and ranges? it seems like we'd want a single field, a non-empty array, that can contain ranges and/or single versions, and that is normatively required to map to at least one existing version of the package.

Author: updated this to use a single key `versions`

| `ranges` | array[string] | Array of VERS-formatted version ranges (e.g., ["vers:npm/>=1.0.0|<2.0.0"]). Required for range-based events. |

Example:
```json
{
  "type": "generalAvailability",
  "effective": "2023-01-01T00:00:00Z",
  "versions": ["1.0.0"],
  "ranges": [],
  "published": "2023-01-01T00:00:00Z",
  "modified": "2023-01-01T00:00:00Z",
  "references": [
    "https://example.com/release-notes"
  ]
}
```

#### endOfDevelopment
*Category: Version Event*

> Indicates when the manufacturer ceases development of a component or service.

Reviewer 2: can you add some examples to help me understand when this would apply?

  "ranges": ["vers:npm/>=1.0.0|<2.0.0"],
  "versions": [],
  "published": "2023-06-01T00:00:00Z",
  "modified": "2023-06-01T00:00:00Z",
}
```

#### endOfSupport
*Category: Version Event*

Indicates when the manufacturer ceases any and all support of a component or service. This point in time marks a transfer of risk from the manufacturer to the consuming organisation or user of the component or service, encompassing all cybersecurity knowledge and known vulnerabilities, with no further assistance provided by the manufacturer.

**Additional Required Fields:**

| Field | Type | Description |
|-------|------|-------------|
| `versions` | array[string] | Array of simple version strings (e.g., ["1.0.0", "1.1.0"]). Required for version-specific events. |
| `ranges` | array[string] | Array of VERS-formatted version ranges (e.g., ["vers:npm/>=1.0.0|<2.0.0"]). Required for range-based events. |

Example:
```json
{
  "type": "endOfSupport",
  "effective": "2024-01-01T00:00:00Z",
  "versions": ["1.0.0", "1.1.0"],
  "ranges": ["vers:npm/>=2.0.0|<3.0.0"],
  "published": "2023-06-01T00:00:00Z",
  "modified": "2023-06-01T00:00:00Z",
}
```

#### endOfGuaranteedSupport
*Category: Version Event*

> Indicates when the manufacturer no longer provides assured services such as technical assistance, user training, repairs, spare parts, and software updates for a component or service. Beyond this period, any support offered is discretionary and may incur extra costs or have restrictions.

Reviewer 2: to me "guaranteed" implies some kind of contract, but open source doesn't have any guarantees - maybe there's a better term to use here?

Author: removed `endOfGuaranteedSupport` after concluding that claims like "guaranteed" can be represented in the support levels.

**Additional Required Fields:**

| Field | Type | Description |
|-------|------|-------------|
| `versions` | array[string] | Array of simple version strings (e.g., ["1.0.0", "1.1.0"]). Required for version-specific events. |
| `ranges` | array[string] | Array of VERS-formatted version ranges (e.g., ["vers:npm/>=1.0.0|<2.0.0"]). Required for range-based events. |

Example:
```json
{
  "type": "endOfGuaranteedSupport",
  "effective": "2024-01-01T00:00:00Z",
  "ranges": ["vers:npm/>=1.0.0|<2.0.0"],
  "versions": [],
  "published": "2023-06-01T00:00:00Z",
  "modified": "2023-06-01T00:00:00Z",
}
```

> #### endOfLife

Reviewer 2: endOfSupport and endOfLife feel equivalent to me. what's the difference?

Author: updated wording to make it clear the difference

*Category: Version Event*

Indicates when the manufacturer stops distribution of a product after its defined useful life and formally notifies users. It marks the conclusion of the product's lifecycle, following a structured EOL process.

**Additional Required Fields:**

| Field | Type | Description |
|-------|------|-------------|
| `versions` | array[string] | Array of simple version strings (e.g., ["1.0.0", "1.1.0"]). Required for version-specific events. |
| `ranges` | array[string] | Array of VERS-formatted version ranges (e.g., ["vers:npm/>=1.0.0|<2.0.0"]). Required for range-based events. |

Example:
```json
{
  "type": "endOfLife",
  "effective": "2024-01-01T00:00:00Z",
  "ranges": ["vers:npm/>=1.0.0|<2.0.0"],
  "versions": [],
  "published": "2023-06-01T00:00:00Z",
  "modified": "2023-06-01T00:00:00Z",
}
```

#### endOfProduction
*Category: Version Event*

> Indicates when the manufacturer stops producing a component, often due to newer versions, unavailable parts, market changes, or strategic shifts. Existing units may still be available in warehouses, distribution channels, and for use, but no new units will be manufactured.

Reviewer 2: this seems like it only covers physical goods, but i thought we decided v1 would just be for software?

Author: removed

**Additional Required Fields:**

| Field | Type | Description |
|-------|------|-------------|
| `versions` | array[string] | Array of simple version strings (e.g., ["1.0.0", "1.1.0"]). Required for version-specific events. |
| `ranges` | array[string] | Array of VERS-formatted version ranges (e.g., ["vers:npm/>=1.0.0|<2.0.0"]). Required for range-based events. |

Example:
```json
{
  "type": "endOfProduction",
  "effective": "2024-01-01T00:00:00Z",
  "ranges": ["vers:npm/>=1.0.0|<2.0.0"],
  "versions": [],
  "published": "2023-06-01T00:00:00Z",
  "modified": "2023-06-01T00:00:00Z",
}
```

#### endOfMarketing
*Category: Version Event*

> Indicates when the manufacturer will cease actively promoting or advertising a component or service. While the component or service may still be available for purchase and support, it will no longer receive active marketing efforts from the manufacturer.

Reviewer 2: why is this interesting to anyone?

**Additional Required Fields:**

| Field | Type | Description |
|-------|------|-------------|
| `versions` | array[string] | Array of simple version strings (e.g., ["1.0.0", "1.1.0"]). Required for version-specific events. |
| `ranges` | array[string] | Array of VERS-formatted version ranges (e.g., ["vers:npm/>=1.0.0|<2.0.0"]). Required for range-based events. |

Example:
```json
{
  "type": "endOfMarketing",
  "effective": "2024-01-01T00:00:00Z",
  "ranges": ["vers:npm/>=1.0.0|<2.0.0"],
  "versions": [],
  "published": "2023-06-01T00:00:00Z",
  "modified": "2023-06-01T00:00:00Z",
}
```

#### supersededBy
*Category: Version Event*

> Indicates when a version of a component is superseded by another version of a component.

Reviewer 2: isn't every version of a component superseded by all future versions? what's an example where this event isn't always implied?

Author: Talked about this in the last meeting: supersededBy is useful in cases where version progression is not implicit (non-semver for example).

| `modified` | string | The time when the event was last modified, as an ISO 8601 formatted timestamp in UTC (ending in "Z"). |

### Event Types

#### released
*Category: Version Event*

Indicates when a component version is released and available for use.

**Additional Required Fields:**

- version

**Additional Optional Fields:**

| Field | Type | Description |
|-------|------|-------------|
| `license` | string | License identifier that summarizes the license as declared in the component's metadata. |

Example:
```json
{
  "type": "released",
  "effective": "2024-01-01T00:00:00Z",
  "version": "1.0.0",
  "license": "MIT",
  "published": "2023-06-01T00:00:00Z",
  "modified": "2023-06-01T00:00:00Z",
}
```

#### endOfDevelopment
*Category: Version Event*

The manufacturer or maintainer ceases work on `Substantial Modifications` for a specific version or version range of a component or service. `Security Fixes` and `Bug Fixes` will continue to be provided for this specific version or version range until `endOfSupport` is declared, but no new features or enhancements will be added.

> > TL;DR: Ceasing Substantial Modifications.

Reviewer 2: it's a bit more than this - generally this means no new features will be added, and no non-security-related bugfixes.

Author: yep correct

> ‘substantial modification’ means a change to the product with digital elements following its placing on the market, which affects the compliance of the product with digital elements with the essential cybersecurity requirements set out in Part I of Annex I or which results in a modification to the intended purpose for which the product with digital elements has been assessed;

Steve was suggesting using the terminology from the CRA for this, I originally had it as `Feature Development`. Im personally having trouble with the wording, so if you have any suggestions, would be very helpful.

Reviewer 2: Let's link here to the CRA's definition of the term; that should be sufficient

**Additional Required Fields:**

- versions
- supportId

Example:
```json
{
  "type": "endOfDevelopment",
  "effective": "2024-01-01T00:00:00Z",
  "versions": [
    {
      "version": "1.0.0"
    },
    {
      "range": "vers:npm/>=1.0.0|<2.0.0"
    }
  ],
  "supportId": "standard",
  "published": "2023-06-01T00:00:00Z",
  "modified": "2023-06-01T00:00:00Z",
}
```

#### endOfSupport
*Category: Version Event*

The manufacturer or maintainer ceases providing `Security Fixes` and `Bug Fixes` for a specific version or version range of a component or service.

The `supportId` field MUST be included and used to specify which support policy is ending, referencing a support policy defined in the definitions section.

> An `endOfSupport` event should only exist when a support definition object exists.

Reviewer 2: ```suggestion
An `endOfSupport` event should only exist when a support definition object exists for the entire indicated version range.
```

> TL;DR: Ceasing Security Fixes and Bug Fixes.

**Additional Required Fields:**

- versions
- supportId

Example:
```json
{
  "type": "endOfSupport",
  "effective": "2024-01-01T00:00:00Z",
  "versions": ["1.0.0"],
  "supportId": "guaranteed",
  "published": "2023-06-01T00:00:00Z",
  "modified": "2023-06-01T00:00:00Z"
}
```

#### endOfLife
*Category: Version Event*

The manufacturer or maintainer formally ceases all work (including `Distribution`, `Substantial Modifications`, `Bug Fixes`, `Security Fixes`, `Documentation`, and `Maintenance`) for a specific version or version range of a component. No further updates, support, or distribution will be provided for this specific version or version range. The component is considered retired.

> TL;DR: Ceasing all work.

**Additional Required Fields:**

- versions

Example:
```json
{
  "type": "endOfLife",
  "effective": "2024-01-01T00:00:00Z",
  "versions": [
    {
      "range": "vers:npm/>=1.0.0|<2.0.0"
    }
  ],
  "published": "2023-06-01T00:00:00Z",
  "modified": "2023-06-01T00:00:00Z",
}
```

### endOfDistribution
*Category: Version Event*

> The manufacturer or maintainer ceases distribution of a specific version or version range of a component or service. This should only be used when the manufacturer has control over the distribution of the component or service.

Reviewer 2: ```suggestion
The manufacturer or maintainer ceases distribution of a specific version or version range of a component or service. This should only be used when the manufacturer has control over the distribution of the component or service - e.g. if the manufacturer publishes to a registry, they likely do not have this control, and thus would not provide this event.
```

The definitions object allows specification of reusable policies and calculations that can be referenced by events.

**Support**

> This is a list of support policies provided for a specific version or version range of a component. Note this SHOULD NOT include third party support policies or options.

Reviewer 2: why should not, if the maintainers have a list of preferred third party options?

Reviewer 2: (ping)

Author: > The support object defines the support policies provided for a specific version or version range of a component. This may include first-party manufacturer support or third-party support options endorsed by the manufacturer.

how does it look?

| Field | Type | Description |
|-------|------|-------------|
| `$schema` | string | URI identifying the [JSON Schema](https://json-schema.org/) document that describes the version of the CLE schema to use (e.g., "https://TODO/cle.v1.0.0.json"). CLE is built on JSON Schema draft 2020-12. |
| `events` | array | Ordered array of Event objects representing the component's lifecycle events. MUST be ordered by ID in descending order (newest events with highest IDs first). |

**Additional Fields:**

| Field | Type | Description |
|-------|------|-------------|
| `versions` | array[object] | Array of version objects. Each object must contain either a `version` field with a simple string or a `range` field with a VERS-formatted version range string. |
| `definitions` | object | Container for reusable policy definitions that can be referenced throughout the document. |

> | `index` | string | URL pointing to the index file that lists all CLE pages for this component. Only required when pagination is used. |

Reviewer 2: ```suggestion
| `index` | string | URL pointing to the index file that lists all CLE pages for this component. Only allowed to be present when pagination is used; and is required in that case. |
```

> | `next` | string | URL pointing to the next CLE page containing newer events (higher event IDs). OPTIONAL - only needed when pagination is used. |

Reviewer 2: ```suggestion
| `next` | string | URL pointing to the next CLE page containing newer events (higher event IDs). When pagination is used, this field is required unless this is the final document. Otherwise, it MUST be absent. |
```

#### endOfMarketing
*Category: Version Event*

The manufacturer or maintainer ceases marketing and promotion of a specific version or version range of a component or service. The component or service may still be available, and existing support policies may remain in effect, but the manufacturer will no longer seek new customers or promote its use.

> > TL;DR: Ceasing marketing and promotion.

Reviewer 2: per the last time we chatted about it, i'm still not sure why this is something anyone cares about.

}
```

#### withdrawn
*Category: Meta Event*

Indicates that a previously published event is being withdrawn or revoked. This is used in a prepend-only event model where events cannot be modified, only new events can be added. When an event is withdrawn, it should be ignored during processing as if it never existed.

**Additional Required Fields:**
| Field | Type | Description |
|-------|------|-------------|
| `eventId` | integer | The ID of the event being withdrawn. Must reference an existing event ID. |

**Additional Optional Fields:**
| Field | Type | Description |
|-------|------|-------------|
| `references` | array[string] | List of URLs to supporting documentation explaining the withdrawal. |
| `reason` | string | Human-readable explanation for why the event is being withdrawn. SHOULD be clear and concise. |

Example:
```json
{
  "id": 9,
  "type": "withdrawn",
  "effective": "2024-02-01T00:00:00Z",
  "published": "2024-02-01T00:00:00Z",

>   "eventId": 3,

Reviewer 2: bikeshed: 
```suggestion
  "targetID": 3,
```

(a pet peeve of mine is capitalizing "id" as "Id", which imo should be reserved for when it's the first word in a sentence about one of Freud's three parts of the mind)
